### PR TITLE
feat: app builder object list

### PR DESCRIPTION
**Pull request by joshbalfour**
# Pull request checklist

**Detail as per issue below (required):**

fixes:  #4618

### CODE REVIEW

File: packages/app-builder/src/components/hooks/objects/use-object.list.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { useQuery } from '@apollo/client'
+import { useIntrospection, dummyQuery } from '../use-introspection'
+
+const getOnlyKey = (obj: any) => obj[Object.keys(obj)[0]]
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/objects/use-object.list.ts:4]:
```suggestion
const getOnlyKey = <T extends any, R extends any>(obj: T): R | undefined => obj[Object.keys(obj)[0]]
```

Is this to get the first entry of the object? `Object.values(obj)[0]` would be a better solution?

File: packages/app-builder/src/components/hooks/use-introspection/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+export const getObjectType = (type: IntrospectionOutputTypeRef): string => {
+  if (isNonNullType(type) || isListType(type)) {
+    return getObjectType(type.ofType)
+  }
+
+  return type.name
+}
+
+export const notEmpty = <TValue>(value: TValue | null | undefined): value is TValue => {
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/use-introspection/helpers.ts:20]:
```suggestion
export const notEmpty = <TValue>(value?: TValue | null): value is TValue => {
```

File: packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import { IntrospectionOutputTypeRef, IntrospectionObjectType } from 'graphql'
+import { flatKind, getObjectType } from './helpers'
+import { LIST, NON_NULL, OBJECT, QueryableField, SCALAR } from './types'
+
+type NestedDict = {
+  [key: string]: NestedDict | undefined
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts:6]:
```suggestion
  [key: string]?: NestedDict
```

File: packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+export const nestedFieldsToString = (nestedFields?: NestedDict) => {
+  if (!nestedFields) {
+    return ''
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts:11]:
```suggestion
    return '{ }'
```

return empty object?

File: packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  if (type.kind === SCALAR) {
+    return undefined
+  }
+
+  if (type.kind === OBJECT) {
+    const objectType = queryableObjectTypes.find(({ name }) => name === type.name)
+    if (objectType) {
+      objectType.fields.forEach((field) => {
+        nestedDict[field.name] = queryableFieldToNestedDict(field.type, queryableObjectTypes)
+      })
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts:41]:
```suggestion
      return objectType.fields.reduce<NestedDist, string>((dict, field) => ({
       ...dict,
       [field.name]: queryableFieldToNestedDict(field.type, queryableObjectTypes),
      }), nestedDict)
```

Might need to reverse the generic types + parameters - I always get reduce params/types the wrong way around

File: packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    }
+  }
+
+  return nestedDict
+}
+
+export const getTopLevelFields = (
+  objectTypes: Array<IntrospectionObjectType>,
+  queryName: string,
+): Array<QueryableField> =>
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts:51]: My old boss used to kill me for not being consistent between `[]` and `Array<>`. I feel like I need to mention it but I really don't want to 😂 I'm not bothered but I can't not mention it now 😱

File: packages/app-builder/src/components/hooks/use-introspection/types.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  IntrospectionOutputTypeRef,
+  IntrospectionNonNullTypeRef,
+  IntrospectionListTypeRef,
+  IntrospectionField,
+} from 'graphql'
+
+export const OBJECT = 'OBJECT'
+export const NON_NULL = 'NON_NULL'
+export const LIST = 'LIST'
+export const SCALAR = 'SCALAR'
```
Comment (bashleigh) [inline packages/app-builder/src/components/hooks/use-introspection/types.ts:12]: Could use an enum here? That way we can group them with a generic type of say `ObjectType` where `ObjectType` is one of these values

Comment (bashleigh) [reply packages/app-builder/src/components/hooks/use-introspection/types.ts:12]: Comes in handy for types like `{[keyof ObjectType]: Function}`

File: packages/app-builder/src/components/ui/user/table.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+const ObjectTableCell = ({ __typename }: any) => {
+  return <span>[{__typename} embedded object]</span>
+}
+
+const Table = ({ typeName, ...props }: TableProps) => {
+  const { data, loading } = useObjectList(typeName)
+  const rows =
+    data &&
+    data.map((row) => {
```
Comment (bashleigh) [inline packages/app-builder/src/components/ui/user/table.tsx:24]: Could use reduce?

File: packages/app-builder/src/components/ui/viewport/index.tsx [hunk 9]
```diff
@@ -56,10 +56,18 @@ const Viewport = ({ children, iframeRef }) => {
...
 
   useEffect(() => {
-    if (pageId) {
-      const page = getPage(pageId)
-      actions.deserialize(page.nodes)
+    let page
+    try {
+      page = getPage(pageId)
+    } catch (e) {
+      //
```
Comment (bashleigh) [inline packages/app-builder/src/components/ui/viewport/index.tsx:63]: 😱

**Review by bashleigh — CHANGES_REQUESTED** [review]
Added quite a few suggestions, guessing most can be ignored for now!

File: packages/app-builder/src/components/hooks/objects/use-object.list.ts (hunk 1 above)
Comment (joshbalfour) [reply packages/app-builder/src/components/hooks/objects/use-object.list.ts:4]: brain fart, you're right 😅

File: packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts (hunk 4 above)
Comment (joshbalfour) [reply packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts:11]: this generates any subfields which should go on to the graphql query, so either 
```
property
```
for a scalar field or
```
property {
  subproperty
}
```
for an object field, so if property has no sub properties you don't add anything, hence the empty string

File: packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts (hunk 5 above)
Comment (joshbalfour) [reply packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts:41]: you got it right :D

File: packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts (hunk 6 above)
Comment (joshbalfour) [reply packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts:51]: i consistently (i hope!) use `Array<any>` but happy to change

**Review by willmcvay — APPROVED** [review]
With Ashleigh's comments LGTM 👍🏻

File: packages/app-builder/src/components/ui/viewport/index.tsx (hunk 9 above)
Comment (joshbalfour) [reply packages/app-builder/src/components/ui/viewport/index.tsx:63]: 😭

File: packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts (hunk 4 above)
Comment (bashleigh) [reply packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts:11]: 👍

**Review by bashleigh — APPROVED** [review]
👍

**Pull request merged**